### ISSUE TITLE
_tweak_price Raises "Loss" Due to Neglible Numerical Error

### ISSUE THREAD

**Issue by nagakingg** [curveresearch/curvesim#224]
### Version Information

curvesim 0.5.0.b1, CPython 3.10.12 on Linux

### What's your issue about?
Simulation is occasionally halted by this assertion failing:
https://github.com/curveresearch/curvesim/blob/1bddb535b692be8d56850abf7b40e4c4b4f9ebb6/curvesim/pool/cryptoswap/pool.py#L383

Exploring this under a number of conditions, I only came across cases where new_virtual_price was 1 less than old_virtual price.

### How can it be fixed?
Either or both:
- Include some tolerance for numerical error in this check
- Raise a warning rather than an exception for this, ideally including the virtual price difference

I think the second option (only) is better to preserve closeness to the original code. But one option would be to always raise a warning, but raise an exception if the difference is greater than some tolerance.

**Comment by chanhosuh** [comment]
if you allow a 1 wei loss, then couldn't this accumulate?

**Comment by nagakingg** [comment]
I looked into this further and we should _not_ ignore this assertion. The reason is obvious in retrospect: especially with high A, the impermanent loss can be greater than would be experenced by a constant-product pool, so XCP can indeed decrease when the pool is imbalanced.

To avoid the AssertionError halting sims, I think we should instead raise a warning and `return` without continuing the `_tweak_price` function. This results in the same functionality without halting sims with the AssertionError.

**Comment by chanhosuh** [comment]
tldr, i recall our discussion concluded it's that we need to also recalculate `tokens`.  Since we recalculate balances to match current `xcp`, integer truncation means when we compute the `xcp` from that, that it's off by 1.  Thus `virtual_price`, which is `10**18 * xcp` / `tokens` is off by 1 potentially.

**Comment by nagakingg** [comment]
Closed by #246

**Comment by nagakingg** [comment]
Closed by #246
